server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -163,40 +164,32 @@ void doRun() {
             return;
         }
 
-        if (isNodeInDuress() == false) {
+        if (nodeDuressTrackers.isNodeInDuress() == false) {
             return;
         }
 
         List<CancellableTask> searchTasks = getTaskByType(SearchTask.class);
         List<CancellableTask> searchShardTasks = getTaskByType(SearchShardTask.class);
-        List<CancellableTask> cancellableTasks = new ArrayList<>();
 
         // Force-refresh usage stats of these tasks before making a cancellation decision.
         taskResourceTrackingService.refreshResourceStats(searchTasks.toArray(new Task[0]));
         taskResourceTrackingService.refreshResourceStats(searchShardTasks.toArray(new Task[0]));
 
-        // Check if increase in heap usage is due to SearchTasks
-        if (HeapUsageTracker.isHeapUsageDominatedBySearch(
-            searchTasks,
-            getSettings().getSearchTaskSettings().getTotalHeapPercentThreshold()
-        )) {
-            cancellableTasks.addAll(searchTasks);
-        }
+        List<TaskCancellation> taskCancellations = new ArrayList<>();
 
-        // Check if increase in heap usage is due to SearchShardTasks
-        if (HeapUsageTracker.isHeapUsageDominatedBySearch(
-            searchShardTasks,
-            getSettings().getSearchShardTaskSettings().getTotalHeapPercentThreshold()
-        )) {
-            cancellableTasks.addAll(searchShardTasks);
-        }
+        taskCancellations = addHeapBasedTaskCancellations(taskCancellations, searchTasks, searchShardTasks);
 
-        // none of the task type is breaching the heap usage thresholds and hence we do not cancel any tasks
-        if (cancellableTasks.isEmpty()) {
-            return;
-        }
+        taskCancellations = addCPUBasedTaskCancellations(taskCancellations, searchTasks, searchShardTasks);
+
+        taskCancellations = addElapsedTimeBasedTaskCancellations(taskCancellations, searchTasks, searchShardTasks);
+
+        // Since these cancellations might be duplicate due to multiple trackers causing cancellation for same task
+        // We need to merge them
+        taskCancellations = mergeTaskCancellations(taskCancellations).stream()
+            .filter(TaskCancellation::isEligibleForCancellation)
+            .collect(Collectors.toList());
```

Comment:
Why not merge and filter in single iteration over all tasks?

Reply:
As discussed offline it can be done but since to do that we will trade off another DS (extra space to achieve it). This is not something which this PR is adding. It was already present hence I think it should not incur any additional overhead